File: solutions/day07/solution.py
from collections import deque

class Solution:
    def avoidFlood(self, rains: list[int]) -> list[int]:
        last = {}
        q = deque()
        res = []

        for i, lake in enumerate(rains):
            if lake != 0:
                if lake in last:
                    found = False
                    temp = deque()
                    while q:
                        j = q.popleft()
                        if not found and j > last[lake]:
                            res[j] = lake
                            found = True
                        else:
                            temp.append(j)
                    if not found:
                        return []
                    q = temp
                res.append(-1)
                last[lake] = i
            else:
                res.append(1)
                q.append(i)

        return res

File: solutions/day07/solution.js
var avoidFlood = function(rains) {
    const last = new Map();
    const q = [];
    const res = [];

    for (let i = 0; i < rains.length; i++) {
        const lake = rains[i];
        if (lake !== 0) {
            if (last.has(lake)) {
                let found = false;
                const temp = [];
                while (q.length > 0) {
                    const j = q.shift();
                    if (!found && j > last.get(lake)) {
                        res[j] = lake;
                        found = true;
                    } else {
                        temp.push(j);
                    }
                }
                if (!found) return [];
                q.push(...temp);
            }
            res.push(-1);
            last.set(lake, i);
        } else {
            res.push(1);
            q.push(i);
        }
    }

    return res;
};
